oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -192,6 +201,61 @@ String createTokenUrlWithScopes() {
     return tokenUrl.toString();
   }
 
+  /**
+   * Gets the universe domain from the GCE metadata server.
+   *
+   * <p>Returns an explicit universe domain if it was provided during credential initialization.
+   *
+   * <p>Returns the {@link Credentials#GOOGLE_DEFAULT_UNIVERSE} if universe domain endpoint is
+   * unavailable or returns an empty string.
+   *
+   * <p>Otherwise, returns universe domain from GCE metadata service.
+   *
+   * <p>Any above value is cached for the credential lifetime.
+   *
+   * @throws IOException if a call to GCE metadata service was unsuccessful. Check if exception
+   *     implements the {@link Retryable} and {@code isRetryable()} will return true if the
+   *     operation may be retried.
+   * @return string representing a universe domain in the format some-domain.xyz
+   */
+  @Override
+  public String getUniverseDomain() throws IOException {
+    if (!isDefaultUniverseDomain()) {
```

Comment:
I think I'd rather break the contract and have the universe also fetched from MDS if the creds are fetched from MDS, for consistency.  WDYT?

Reply:
i was hesitant as well :) I think one key scenario for explicit credential is testing, and it might be relevant for GCE credential as well.

Reply:
Per Slava, we want to maintain the contract here

Reply:
If the rule is to return the universe domain set by the user for testing, we should do that consistently whether or not it is the 'default'.  To keep the contract here you would need to return GDU if that was set in the construction of the object.  Otherwise, I can't force a GDU value in my tests.

To do that, you might do something like update the builder for this class to set a "user set universe domain" flag when users call `setUniverseDomain()` and consult that bit here.

Reply:
Oh you are right... somehow I fixed on testing the custom universe domain case.

Reply:
Added the isExplicitUniverseDomain